torch/_inductor/decomposition.py

```diff
@@ -269,6 +269,62 @@ def angle(x):
         return ret + nan
 
 
+@register_decomposition([aten.add])
+def add(x, y, **kwargs):
+    x_is_complex_tensor = torch.is_tensor(x) and x.is_complex()
+    y_is_complex_tensor = torch.is_tensor(y) and y.is_complex()
+    if not x_is_complex_tensor and not y_is_complex_tensor:
+        return NotImplemented
+    r = y.real
+    i = y.imag if y_is_complex_tensor else 0.0
+    alpha = kwargs.get("alpha")
+    if alpha is not None:
+        r *= alpha
+        i *= alpha
+    r = x.real + r
+    if x_is_complex_tensor:
+        i = x.imag + i
+    complex_type = x.dtype if x_is_complex_tensor else y.dtype
+    return (
+        torch.where(
+            torch.arange(2, device=x.device, dtype=torch.uint8) == 0,
+            r.unsqueeze(-1),
+            i.unsqueeze(-1),
+        )
+        .view(complex_type)
+        .squeeze(-1)
+    )
```

Comment:
For add in particular I think you can just do:
```
return (x.view(torch.float32)+x.view(torch.float32)).view(complex_type)
```
Since you do the same thing for both .real and .complex.

Reply:
This is cool. Thanks for the suggestion. However, the generated code still have `aten.view` calls and CSE cannot be done across kenerls. I guess the goal here is to optimize away `aten.view`?

   ```
 with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        # Source Nodes: [Z], Original ATen: [aten.add]
        buf0 = aten.view(arg0_1, torch.float32)
        del arg0_1
        buf1 = buf0
        del buf0
        # Source Nodes: [Z], Original ATen: [aten.add]
        buf2 = aten.view(arg1_1, torch.float32)
        del arg1_1
        buf3 = buf2
        del buf2
        buf4 = buf1; del buf1  # reuse
        # Source Nodes: [Z], Original ATen: [aten.add]
        stream0 = get_cuda_stream(0)
        triton_poi_fused_add_0.run(buf4, buf3, 2000000, grid=grid(2000000), stream=stream0)
        del buf3
        # Source Nodes: [Z], Original ATen: [aten.add]
        buf5 = aten.view(buf4, torch.complex64)
        del buf4
        buf6 = buf5
        del buf5
        return (buf6, )
```

Reply:
Yeah, fixing the view problems requires teaching inductor how to do complex views.

The scalar case would need to be handled differently.  You could first convert the scalar to complex.

Reply:
@jansel I think an easier thing to do would just be to pattern-match away redundant view calls right?

Reply:
Where would you do that? BTW, are they redundant? I thought without the view calls, the type of the tensors wouldn't be correct.

Reply:
In this case, they're not redundant. But if you have two adds in a row, the views in the way would prevent fusion.

So for a single complex add, it would be 
```
x.view(torch.float64)
x + x
x.view(torch.complex64)
```

Then for two complex adds, it would be
```
x.view(torch.float64)
x + x
x.view(torch.complex64)
x.view(torch.float64)
x + x
x.view(torch.complex64)
```

Reply:
There is the additional problem in that these views are getting mapped to fallback kernels, which don't properly encode the aliasing relationship.  So inductor may try to reuse the memory for something else, which is a correctness issue.

Reply:
I’m working on a new inductor IR node that handles complex/float views to avoid calling into the runtime. According to recent discussion, it looks like the triton community doesn’t like supporting complex natively there. But do you see a way to handle the views without using a triton kernel? Once the views are lowered, inductor will fuse the lowered code into the same triton kernel, which exposes complex64 to triton.